packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebView.java

```diff
@@ -24,14 +30,108 @@
+
+    InputAwareWebView(Context context, View flutterView) {
+      super(context);
+      this.flutterView = flutterView;
+    }
+
+    private View threadedInputConnectionProxyView;
+
+    private ThreadedInputConnectionProxyAdapterView proxyAdapterView;
+
+    @Override
+    public InputConnection onCreateInputConnection(EditorInfo outAttrs) {
+      if (proxyAdapterView == null) {
+        // No proxy adapter set, so we use the default implementation.
+        return super.onCreateInputConnection(outAttrs);
+      }
+
+      if (!proxyAdapterView.isTriggerDelayed()) {
+        // Currently on the IME thread. Delegate to the superclass.
+        return super.onCreateInputConnection(outAttrs);
+      }
+
+      InputConnection result = super.onCreateInputConnection(outAttrs);
+      if (result != null) {
+        // This should never happen.
+        Log.wtf(TAG, "Failed unexpectedly creating a webview input connection.");
+      }
+
+      final InputMethodManager imm =
+          (InputMethodManager) getContext().getSystemService(INPUT_METHOD_SERVICE);
+      proxyAdapterView.requestFocus();
+      final View containerView = this;
+
+      // This is the crucial trick that gets the InputConnection creation to happen on the correct
+      // thread. https://cs.chromium.org/chromium/src/content/public/android/java/src/org/chromium/content/browser/input/ThreadedInputConnectionFactory.java?l=169.
+      this.post(
+          new Runnable() {
+            @Override
+            public void run() {
+              // This is a hack to make InputMethodManager believe that the proxy view now has a focus.
+              // As a result, InputMethodManager will think that mProxyView is focused, and will call
+              // getHandler() of the view when creating input connection.
+
+              // Step 1: Set proxyAdapterView as InputMethodManager#mNextServedView. This does not
+              // affect the real window focus.
+              proxyAdapterView.onWindowFocusChanged(true);
+
+              // Step 2: Have InputMethodManager focus in on containerView. As a result, IMM will call
+              // onCreateInputConnection() on proxyAdapterView on the same thread as
+              // proxyAdapterView.getHandler(). It will also call subsequent InputConnection methods on
+              // this IME thread.
+              imm.isActive(containerView);
+            }
+          });
+      return null;
+    }
+
+    @Override
+    public boolean checkInputConnectionProxy(final View view) {
+      View previousProxy = threadedInputConnectionProxyView;
+      threadedInputConnectionProxyView = view;
+      if (previousProxy != view) {
+        proxyAdapterView =
+            new ThreadedInputConnectionProxyAdapterView(
+                /*containerView=*/ flutterView,
+                /*targetView=*/ view,
+                /*imeHandler=*/ view.getHandler());
+        final View container = this;
+        proxyAdapterView.requestFocus();
+        post(
```

Comment:
nit: add a comment explaining that checkInputConnectionProxy for the proxy view happened after WebView#onCreateInputConnection, so we're triggering another onCreateInputConnectionProxy to start the dance to set up a proxy view.

Actually - I wonder if we can just start the focus/isActive dance directly here without calling `imm.restartInput`.